scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/TaskAction.java

```diff
@@ -41,97 +64,103 @@ protected ReadNewWindowTaskAction(WorkerConfiguration workerConfiguration, Task
-
-            CompletableFuture<Reader> readerFuture = waitFuture.thenCompose(w -> workerConfiguration.cql.createReader(task));
-            CompletableFuture<TaskAction> taskActionFuture = readerFuture
-                    .thenApply(reader -> new ReadChangeTaskAction(workerConfiguration, task, reader, tryAttempt));
-            return FutureUtils.thenComposeExceptionally(taskActionFuture, ex -> {
-                // Exception occured while starting up the reader. Retry by starting
-                // this TaskAction once again.
-                long backoffTime = workerConfiguration.workerRetryBackoff.getRetryBackoffTimeMs(tryAttempt);
-                logger.atSevere().withCause(ex).log("Error while starting reading next window. Task: %s. " +
-                        "Task state: %s. Will retry after backoff (%d ms).", task.id, task.state, backoffTime);
-                return workerConfiguration.delayedFutureService.delayedFuture(backoffTime)
-                        .thenApply(t -> new ReadNewWindowTaskAction(workerConfiguration, task, tryAttempt + 1));
-            });
+            try {
+                CompletableFuture<Void> waitFuture = waitForWindow();
+
+                CompletableFuture<Reader> readerFuture = waitFuture
+                        .thenCompose(w -> workerConfiguration.cql.createReader(task));
+                CompletableFuture<TaskAction> taskActionFuture = readerFuture.thenApply(
+                        reader -> new ReadChangeTaskAction(workerConfiguration, task, reader, tryAttempt, task.state));
+                return FutureUtils.thenComposeExceptionally(taskActionFuture, this::onException);
+            } catch (Throwable ex) {
+                return onException(ex);
+            }
         }
 
         private CompletableFuture<Void> waitForWindow() {
             Date end = task.state.getWindowEndTimestamp().toDate();
             Date now = new Date();
             long toWait = end.getTime() - now.getTime() + workerConfiguration.confidenceWindowSizeMs;
             if (toWait > 0) {
-                return workerConfiguration.delayedFutureService.delayedFuture(toWait);
+                return delay(toWait);
             }
             return CompletableFuture.completedFuture(null);
         }
     }
 
     private static class ReadChangeTaskAction extends TaskAction {
-        private final WorkerConfiguration workerConfiguration;
-        private final Task task;
         private final Reader reader;
         private final int tryAttempt;
-
-        public ReadChangeTaskAction(WorkerConfiguration workerConfiguration, Task task, Reader reader, int tryAttempt) {
-            this.workerConfiguration = Preconditions.checkNotNull(workerConfiguration);
-            this.task = Preconditions.checkNotNull(task);
+        private final TaskState newState;
+        
+        public ReadChangeTaskAction(WorkerConfiguration workerConfiguration, Task task, Reader reader, int tryAttempt, TaskState newState) {
```

Comment:
This change adds a bit of confusion: there are now two `TaskState`:
1. `task.state`
2. `newState`

Both `TaskState` are used - `newState` is passed to `transport` in `run()`, while `task.state` is used in `onException()` to queue a retry reading of window.

In the upcoming "event" abstraction, I believe there are only two cases:
1. `task.state != null`, `newState == null` - `newState` indicates `setState()` in transport not to be called.
2. `task.state == newState` - `newState` indicates `setState()` to be called.

Essentially `newState` acts as a boolean value. I don't see why anyone would ever make a `Consumer` that would return `newState` that is `newState != null && newState != task.state`. Such a `Consumer` would wreak mayhem in the library - I personally could not reason about whether retries would be correct in such a case.

Reply:
In a sense, yes, it is a boolean. This is internal mechanism, and the main reason it looks like this is to fit into the existing code flow (reasonably). But there is another point: Assuming task state gets more semantically rich, it might be prudent for the mechanism that wants to operate on a batch sequence level of granularity, to actually include more/different data here. 
This is why it is not just a boolean.

Reply:
I think it would be better not to expose the TaskState if it's not needed at the moment. If it's needed in the future then we will expose it but until then it's better to keep it hidden.

Reply:
Uhm, it is already an integral part of the `WorkerTransport` interface. I wholly agree though that it is somewhat overexposed. It should be an interface + serialization, so the user can store/provide it, but not really read it. 
As for the state-consuming interface here, it is not public. Package-local internal dispatching.